File: anr-watchdog/src/test/java/com/github/anrwatchdog/ANRWatchDogTest.java
package com.github.anrwatchdog;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE)
public class ANRWatchDogTest {

    @Test
    public void testDefaultConstructor() {
        ANRWatchDog watchDog = new ANRWatchDog();

        assertEquals(5000, watchDog.getTimeoutInterval());
    }

    @Test
    public void testCustomTimeoutConstructor() {
        int customTimeout = 10000;

        ANRWatchDog watchDog = new ANRWatchDog(customTimeout);

        assertEquals(customTimeout, watchDog.getTimeoutInterval());
    }

    @Test
    public void testSetANRListener_returnsItself() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setANRListener(error -> {});

        assertSame(watchDog, result);
    }

    @Test
    public void testSetANRListener_nullResetsToDefault() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setANRListener(null);

        assertSame(watchDog, result);
    }

    @Test
    public void testSetANRInterceptor_returnsItself() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setANRInterceptor(duration -> 0);

        assertSame(watchDog, result);
    }

    @Test
    public void testSetANRInterceptor_nullResetsToDefault() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setANRInterceptor(null);

        assertSame(watchDog, result);
    }

    @Test
    public void testSetInterruptionListener_returnsItself() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setInterruptionListener(exception -> {});

        assertSame(watchDog, result);
    }

    @Test
    public void testSetInterruptionListener_nullResetsToDefault() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setInterruptionListener(null);

        assertSame(watchDog, result);
    }

    @Test
    public void testSetReportThreadNamePrefix_returnsItself() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setReportThreadNamePrefix("MyApp:");

        assertSame(watchDog, result);
    }

    @Test
    public void testSetReportThreadNamePrefix_nullSetsEmptyPrefix() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setReportThreadNamePrefix(null);

        assertSame(watchDog, result);
    }

    @Test
    public void testSetReportMainThreadOnly_returnsItself() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setReportMainThreadOnly();

        assertSame(watchDog, result);
    }

    @Test
    public void testSetReportAllThreads_returnsItself() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setReportAllThreads();

        assertSame(watchDog, result);
    }

    @Test
    public void testSetLogThreadsWithoutStackTrace_returnsItself() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setLogThreadsWithoutStackTrace(true);

        assertSame(watchDog, result);
    }

    @Test
    public void testSetIgnoreDebugger_returnsItself() {
        ANRWatchDog watchDog = new ANRWatchDog();

        ANRWatchDog result = watchDog.setIgnoreDebugger(true);

        assertSame(watchDog, result);
    }

    @Test
    public void testFluentApi_chainingWorks() {
        ANRWatchDog watchDog = new ANRWatchDog(3000)
                .setANRListener(error -> {})
                .setANRInterceptor(duration -> 0)
                .setInterruptionListener(exception -> {})
                .setReportThreadNamePrefix("Test:")
                .setLogThreadsWithoutStackTrace(true)
                .setIgnoreDebugger(true);

        assertNotNull(watchDog);
        assertEquals(3000, watchDog.getTimeoutInterval());
    }

    @Test
    public void testIsThread() {
        ANRWatchDog watchDog = new ANRWatchDog();

        assertTrue(watchDog instanceof Thread);
    }

    @Test
    public void testDifferentTimeouts() {
        ANRWatchDog watchDog1 = new ANRWatchDog(1000);
        ANRWatchDog watchDog2 = new ANRWatchDog(2000);
        ANRWatchDog watchDog3 = new ANRWatchDog(30000);

        assertEquals(1000, watchDog1.getTimeoutInterval());
        assertEquals(2000, watchDog2.getTimeoutInterval());
        assertEquals(30000, watchDog3.getTimeoutInterval());
    }

    @Test
    public void testCustomListener_receivesError() {
        AtomicBoolean listenerCalled = new AtomicBoolean(false);
        AtomicLong receivedDuration = new AtomicLong(0);

        ANRWatchDog.ANRListener listener = error -> {
            listenerCalled.set(true);
            receivedDuration.set(error.duration);
        };

        // Just test the listener can be set without errors
        ANRWatchDog watchDog = new ANRWatchDog().setANRListener(listener);

        assertNotNull(watchDog);
    }

    @Test
    public void testCustomInterceptor_canBeSet() {
        ANRWatchDog.ANRInterceptor interceptor = duration -> {
            if (duration < 5000) {
                return 5000 - duration;
            }
            return 0;
        };

        ANRWatchDog watchDog = new ANRWatchDog().setANRInterceptor(interceptor);

        assertNotNull(watchDog);
    }
}

File: anr-watchdog/src/test/kotlin/com/github/anrwatchdog/ANRWatchDogTest.kt
package com.github.anrwatchdog

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE)
class ANRWatchDogTest {

    @Test
    fun `default constructor uses 5000ms timeout`() {
        val watchDog = ANRWatchDog()

        assertEquals(5000, watchDog.getTimeoutInterval())
    }

    @Test
    fun `custom timeout constructor`() {
        val customTimeout = 10000

        val watchDog = ANRWatchDog(customTimeout)

        assertEquals(customTimeout, watchDog.getTimeoutInterval())
    }

    @Test
    fun `setANRListener returns itself for chaining`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setANRListener { }

        assertSame(watchDog, result)
    }

    @Test
    fun `setANRListener with null resets to default`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setANRListener(null)

        assertSame(watchDog, result)
    }

    @Test
    fun `setANRInterceptor returns itself for chaining`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setANRInterceptor { 0 }

        assertSame(watchDog, result)
    }

    @Test
    fun `setANRInterceptor with null resets to default`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setANRInterceptor(null)

        assertSame(watchDog, result)
    }

    @Test
    fun `setInterruptionListener returns itself for chaining`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setInterruptionListener { }

        assertSame(watchDog, result)
    }

    @Test
    fun `setInterruptionListener with null resets to default`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setInterruptionListener(null)

        assertSame(watchDog, result)
    }

    @Test
    fun `setReportThreadNamePrefix returns itself for chaining`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setReportThreadNamePrefix("MyApp:")

        assertSame(watchDog, result)
    }

    @Test
    fun `setReportThreadNamePrefix with null sets empty prefix`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setReportThreadNamePrefix(null)

        assertSame(watchDog, result)
    }

    @Test
    fun `setReportMainThreadOnly returns itself for chaining`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setReportMainThreadOnly()

        assertSame(watchDog, result)
    }

    @Test
    fun `setReportAllThreads returns itself for chaining`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setReportAllThreads()

        assertSame(watchDog, result)
    }

    @Test
    fun `setLogThreadsWithoutStackTrace returns itself for chaining`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setLogThreadsWithoutStackTrace(true)

        assertSame(watchDog, result)
    }

    @Test
    fun `setIgnoreDebugger returns itself for chaining`() {
        val watchDog = ANRWatchDog()

        val result = watchDog.setIgnoreDebugger(true)

        assertSame(watchDog, result)
    }

    @Test
    fun `fluent API chaining works`() {
        val watchDog = ANRWatchDog(3000)
            .setANRListener { }
            .setANRInterceptor { 0 }
            .setInterruptionListener { }
            .setReportThreadNamePrefix("Test:")
            .setLogThreadsWithoutStackTrace(true)
            .setIgnoreDebugger(true)

        assertNotNull(watchDog)
        assertEquals(3000, watchDog.getTimeoutInterval())
    }

    @Test
    fun `ANRWatchDog is a Thread`() {
        val watchDog = ANRWatchDog()

        assertTrue(watchDog is Thread)
    }

    @Test
    fun `different timeouts are stored correctly`() {
        val watchDog1 = ANRWatchDog(1000)
        val watchDog2 = ANRWatchDog(2000)
        val watchDog3 = ANRWatchDog(30000)

        assertEquals(1000, watchDog1.getTimeoutInterval())
        assertEquals(2000, watchDog2.getTimeoutInterval())
        assertEquals(30000, watchDog3.getTimeoutInterval())
    }

    @Test
    fun `custom listener can be set`() {
        var listenerCalled = false

        val listener = ANRWatchDog.ANRListener {
            listenerCalled = true
        }

        val watchDog = ANRWatchDog().setANRListener(listener)

        assertNotNull(watchDog)
    }

    @Test
    fun `custom interceptor can be set`() {
        val interceptor = ANRWatchDog.ANRInterceptor { duration ->
            if (duration < 5000) 5000 - duration else 0
        }

        val watchDog = ANRWatchDog().setANRInterceptor(interceptor)

        assertNotNull(watchDog)
    }
}
